src/plots/mapbox/mapbox.js

```diff
@@ -0,0 +1,377 @@
+            // to avoid 'lost event' errors
+            self.traceHash = {};
+
+            self.updateData(calcData);
+            self.updateLayout(fullLayout);
+
+            self.resolveOnRender(resolve);
+        });
+    }
+    else {
+        self.updateData(calcData);
+        self.updateLayout(fullLayout);
+
+        self.resolveOnRender(resolve);
+    }
+};
+
+proto.updateData = function(calcData) {
+    var traceHash = this.traceHash;
+
+    var traceObj, trace, i, j;
+
+    // update or create trace objects
+    for(i = 0; i < calcData.length; i++) {
+        var calcTrace = calcData[i];
+
+        trace = calcTrace[0].trace;
+        traceObj = traceHash[trace.uid];
+
+        if(traceObj) traceObj.update(calcTrace);
+        else {
+            traceHash[trace.uid] = trace._module.plot(this, calcTrace);
+        }
+    }
+
+    // remove empty trace objects
+    var ids = Object.keys(traceHash);
+    id_loop:
+    for(i = 0; i < ids.length; i++) {
+        var id = ids[i];
+
+        for(j = 0; j < calcData.length; j++) {
+            trace = calcData[j][0].trace;
+
+            if(id === trace.uid) continue id_loop;
+        }
+
+        traceObj = traceHash[id];
+        traceObj.dispose();
+        delete traceHash[id];
+    }
+};
+
+proto.updateLayout = function(fullLayout) {
+    var map = this.map,
+        opts = this.opts;
+
+    map.setCenter(convertCenter(opts.center));
+    map.setZoom(opts.zoom);
+    map.setBearing(opts.bearing);
+    map.setPitch(opts.pitch);
+
+    this.updateLayers();
+    this.updateFramework(fullLayout);
+    this.map.resize();
+};
+
+proto.resolveOnRender = function(resolve) {
+    var map = this.map;
+
+    map.on('render', function onRender() {
+        if(map.loaded()) {
+            map.off('render', onRender);
+            resolve();
+        }
+    });
+};
+
+proto.rejectOnError = function(reject) {
+    var map = this.map;
+
+    function handler() {
+        reject(new Error(constants.mapOnErrorMsg));
+    }
+
+    map.once('error', handler);
+    map.once('style.error', handler);
+    map.once('source.error', handler);
+    map.once('tile.error', handler);
+    map.once('layer.error', handler);
+};
+
+proto.createFramework = function(fullLayout) {
+    var self = this;
+
+    var div = self.div = document.createElement('div');
+
+    div.id = self.uid;
+    div.style.position = 'absolute';
+
+    self.container.appendChild(div);
+
+    // create mock x/y axes for hover routine
+
+    self.xaxis = {
+        _id: 'x',
+        c2p: function(v) { return self.project(v).x; }
+    };
+
+    self.yaxis = {
+        _id: 'y',
+        c2p: function(v) { return self.project(v).y; }
+    };
+
+    self.updateFramework(fullLayout);
+};
+
+proto.updateFramework = function(fullLayout) {
+    var domain = fullLayout[this.id].domain,
+        size = fullLayout._size;
+
+    var style = this.div.style;
+
+    // TODO Is this correct? It seems to get the map zoom level wrong?
```

Comment:
I'm not sure. I need to look in the mapbox internals to see how they convert viewport dimension to canvas coords. It is different than in our `geo` code. 

I'm _ok_ with the current behavior for this beta mode. But it is subject to change.